spirhunt/spirhunt.6
.\" This file is free software, distributed under the BSD license.
.Dd December 30, 1993
.Dt SPIRHUNT 6
.Os
.Sh NAME
.Nm spirhunt
.Nd space combat game
.Sh SYNOPSIS
.Nm
.Oo
.Op Fl a
.Ar file
.Oc
.Sh DESCRIPTION
.Nm
is a game of space glory and war.
Below is a summary of commands.
For complete documentation, see
.Em Space Pirate Hunt
by Eric Allman.
.Pp
In general, throughout the game, if you forget what is appropriate
the game will tell you what it expects if you just type in a question mark.
.Sh AUTHOR
.An Eric Allman
.Sh COMMAND SUMMARY
.Bl -item -compact
.It
.Ic ca Ns pture
.It
.Ic cl Ns oak
.Ic u Ns p/ Ns Ic d Ns own
.It
.Ic da Ns mages
.It
.Ic destruct
.It
.Ic do Ns ck
.It
.Ic help
.It
.Ic i Ns mpulse qqss
.It
.Ic l Ns rscan
.It
.Ic m Ns ove qqss
.It
.Ic p Ns lasers
.It
.Ic t Ns orpedo
.It
.Ic ram No dest
.It
.Ic r Ns est time
.It
.Ic sh Ns ields
.Ic u Ns p/ Ns Ic d Ns own
.It
.Ic s Ns rscan
.Op Ic y Ns es/ Ic n Ns o
.It
.Ic quit y Ns es/ Ns Ic n Ns o
.It
.Ic u Ns ndock
.It
.Ic v Ns isual course
.It
.Ic w Ns arp warp_factor
.El


doc/NEWS
NEWS for bsd-games and bsd-games-non-free
=========================================

This file contains a summary of changes in each version of bsd-games
and bsd-games-non-free after 1.3.  For fuller details, see ChangeLog
and ChangeLog.0.

As well as reading this summary, it is a good idea to reread the
INSTALL file whenever installing a new version, to check for any
changes in installation instructions.  Packagers should also reread
the PACKAGING file each version (or do a diff between the old and new
versions of the package and read that).

Version 2.17 (2005-02-18)
============

* Updated from NetBSD CVS of 2005-02-18.

* Some bugs fixed (in particular, should now build with current GCC 4
CVS), some backlogged.

Version 2.16 (2004-02-13)
============

* Updated from NetBSD CVS of 2004-02-13.  Includes a new game, dab
(Dots and Boxes), written in C++.

* Supplementary documentation for trek and rogue now installed
automatically (in a configurable directory).

* Some changes for portability to Cygwin, thanks to David A Bagley
<bagleyd@tux.org>; more portability issues remain.

* Various improvements to morse from OpenBSD.

* Some bugs fixed, some backlogged.

Version 2.15 (2003-12-18)
============

* Updated from NetBSD CVS of 2003-12-17.  This includes:

  + Advertising clause removed from licence of almost all
  UCB-copyright code.

  + Hunt licence clarified.

* Various bugs fixed, though some bug reports are still backlogged.

* Glibc 2.3 or later now required, at least for hunt.

* Directory for fortune files individually configurable, and default
has changed to match that used in NetBSD.

Version 2.14 (2003-04-01)
============

* Hack is now BSD-licensed
<http://www.cwi.nl/~aeb/games/hack/hack.html>, so moves from
bsd-games-non-free to bsd-games.

* Updated from NetBSD CVS of 2003-03-29.  This includes:

  + Security fix for rogue (in bsd-games-non-free); see NetBSD
  Security Advisory 2002-021.

  + Optional support for large numbers in factor using libcrypto from
  OpenSSL.  This can handle larger numbers than GNU factor (which is
  limited to 64-bit numbers) but not that much larger efficiently, as
  the algorithm used (Pollard rho) is O(n^{1/4}) in the number
  factored.

* Some bugs fixed, though a backlog of bug reports remains.

Version 2.13 (2001-08-30)
============

* Warning now present in README about the inclusion of cryptographic
software (caesar and rot13).

* Various minor improvements and bugfixes, some from NetBSD, some from
OpenBSD, some from other contributors.  Current with NetBSD CVS of
2001-08-29.

* Some fixes to avoid some warnings when compiling with current glibc
and GCC 3.

* May support GNU Hurd (untested).

Version 2.12 (2000-09-25)
============

* Improvements and bugfixes to battlestar, mainly from OpenBSD.
Though there are still plenty of known bugs, it should be much harder
to crash.

* Should compile with glibc 2.2 (tested with 2.1.94).

* Other minor changes and bugfixes; current with NetBSD CVS of
2000-09-25.

Version 2.11 (2000-04-19)
============

* bsd-games-non-free now includes hack 1.0.3 by Andries Brouwer; this
may be of interest to nethack players wishing to play an ancestral
game.

* Various minor changes and bugfixes; current with NetBSD CVS of
2000-04-19.

Version 2.10 (2000-02-12)
============

* Avoid world-readable characs file in phantasia.  Note the new
configuration question which will need an appropriate answer,
especially if packaging for a distribution.

* Various minor bugfixes.

* Essentially all relevant changes, except to hunt (which will
eventually be replaced by the OpenBSD version), now merged back into
NetBSD.

* Some minor changes from NetBSD CVS of 2000-02-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.9 (1999-11-12)
===========

* Fixes various minor bugs reported recently.

* Some minor changes from NetBSD CVS of 1999-11-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.8 (1999-09-26)
===========

* Robots scorefile format has changed - remove or rename your old
score file when upgrading to this version.

* Various minor changes and bug-fixes (some from OpenBSD).

* Snake substantially cleaned up, and now uses the native curses
interface (rather than termcap emulation in ncurses); backgammon and
tetris still to be converted.

* New "game": wtf.

* Updated from NetBSD CVS of 1999-09-25; many more changes merged back
into NetBSD.

* All constant data files are now architecture independent (installed
under /usr/share by default).

* bsd-games is no longer tested by me with libc older than glibc
2.1.2; fixes for compilation with older libc will still be accepted.

* Details of the copyrights and licence terms applicable to these
games, as best as I can determine them, now appear in COPYING.

* bsd-games no longer comes with its own word list for boggle and
hangman; you need to provide one at compile time (for boggle) and at
run time (for hangman).  The copyright status of the one formerly
included was not entirely clear.  The word list used by hangman can
also now be specified at run time with the new "-d" option.

* bsd-games now has a testsuite (for programs that can be tested
non-interactively only).

Version 2.7 (1999-04-04)
===========

* Bug fixes.

* Updated from NetBSD-current of 1999-03-26; more changes merged back
into NetBSD.

* Other minor changes.

Version 2.6 (1998-12-20)
===========

* Buffer overrun fixes in sail (not thought exploitable).

* Fixes possible score file corruption if fds 0, 1, 2 aren't open at
game startup.

* Updated from NetBSD-current of 1998-12-07 (morse now has a decoding
option).

* Now builds cleanly with -Wwrite-strings.

* Other minor changes.

* Some users of this package may be interested in my "nppt" package,
which provides an alternative implementation of ppt(6) with additional
features such as Encapsulated PostScript output.  (This package

spirhunt/spirhunt.c
// Copyright (c) 1980 The Regents of the University of California.
// This file is free software, distributed under the BSD license.

#include "spirhunt.h"
#include "getpar.h"
#include <sys/uio.h>
#include <sys/stat.h>

//{{{ Local prototypes -------------------------------------------------

static void setup (void);
static void play (void);
static uint16_t sum_save_array (void);
static bool verify_locations (void);
static bool restore (void);

//}}}-------------------------------------------------------------------
//{{{ Global variables

//{{{2 DeviceName
const char DeviceName [NDEV][16] = {
    "warp drive",
    "near scanners",
    "far scanners",
    "plasers",
    "torpedoes",
    "impulse engines",
    "shield control",
    "computer",
    "radio",
    "life support",
    "cloak",
    "transporter",
    "shuttlecraft"
};
//}}}2
//{{{2 Systemname
const char Systemname[NINHAB][16] = {
    "ERROR",
    "Talos IV",
    "Rigel III",
    "Deneb VII",
    "Canopus V",
    "Icarus I",
    "Prometheus II",
    "Omega VII",
    "Elysium I",
    "Scalos IV",
    "Procyon IV",
    "Arachnid I",
    "Argo VIII",
    "Triad III",
    "Echo IV",
    "Nimrod III",
    "Nemisis IV",
    "Centarurus I",
    "Kronos III",
    "Spectros V",
    "Beta III",
    "Tranguli VI",
    "Pyris III",
    "Triachus",
    "Marcus XII",
    "Kaland",
    "Ardana",
    "Stratos",
    "Eden",
    "Arrikis",
    "Eridani IV",
    "Exo III"
};
//}}}2
//{{{2 Param

const struct Param_struct Param = {
    .crew	= 387,
    .brigfree	= 400,
    .energy	= 5000,
    .energylow	= 1000,
    .shield	= 1500,
    .shupengy	= 40,
    .stopengy	= 50,
    .cloakenergy= 1000,
    .pirate_crew= 200,
    .piratepwr	= 350,
    .bases	= MAXBASES,
    .pirates	= NQUADS*NQUADS/2,
    .torped	= 10,
    .warptime	= 10,
    .moveprob = {
	[PM_EA]	= 35,
	[PM_EB]	= 40,
	[PM_LA]	= 30,
	[PM_LB]	= 0,
	[PM_OA]	= 40,
	[PM_OB]	= 45
    },
    .date	= 2000,
    .time	= 14,
    .reserves	= 10,
    .resource	= (NQUADS*NQUADS/2)*14,
    .dockfac	= 0.5,
    .hitfac	= 0.5,
    .plasfac	= 0.8,
    .regenfac	= 0.2,
    .damfac = {
	0.5, 0.5, 0.5, 0.5, 0.5,
	0.5, 0.5, 0.5, 0.5, 0.5,
	0.5, 0.5, 0.5
    },
    .movefac = {
	[PM_EA]	= -0.12,
	[PM_EB]	= 0.075,
	[PM_LA]	= 0.25,
	[PM_LB]	= 0.0,
	[PM_OA]	= -0.05,
	[PM_OB]	= .09
    },
    .navigcrud	= { 1.50, 0.75 },
    .eventdly = {
	[E_SNOVA] = 0.5,
	[E_LRTB]  =25.0,
	[E_PATSB] = 1.0,
	[E_PDESB] = 3.0,
	[E_ISSUE] = 1.0,
	[E_SNAP]  = 0.5,
	[E_ENSLV] = 0.5,
	[E_REPRO] = 2.0
    }
};
//}}}2

struct event	Event [MAXEVENTS] = {};	// dynamic event list; one entry per pending event
struct quad	Quad [NQUADS][NQUADS] = {};

struct Ship_struct	Ship = {};
struct Game_struct	Game = {};
struct Move_struct	Move = {};
struct Now_struct	Now = {};
struct Etc_struct	Etc = {};

struct Snapshot_struct	Snapshot = {};

//}}}-------------------------------------------------------------------

int main (void)
{
    srandrand();
    if (!restore()) {
	printf ("\n" BOLD_ON "   * * *   SPACE PIRATE HUNT   * * *" BOLD_OFF "\n\n");
	setup();
    }
    play();
    return EXIT_SUCCESS;
}

//{{{ setup ------------------------------------------------------------

static void setup (void)
{
    Now.resource = Param.resource;
    Ship.energy = Param.energy;
    Ship.torped = Param.torped;
    Ship.shield = Param.shield;
    Ship.reserves = Param.reserves;
    Ship.crew = Param.crew;
    Ship.brigfree = Param.brigfree;
    Ship.shldup = true;
    Ship.cond = GREEN;
    Ship.warp = 50;
    Now.date = Param.date + nrand (Param.date/20);
    xsched(E_SNOVA, 1, 0, 0, 0);
    xsched(E_LRTB, Param.pirates, 0, 0, 0);
    xsched(E_PATSB, 1, 0, 0, 0);
    xsched(E_ISSUE, 1, 0, 0, 0);
    xsched(E_SNAP, 1, 0, 0, 0);
    Ship.sect.x = nrand(NSECTS);
    Ship.sect.y = nrand(NSECTS);

    // setup stars
    uint8_t inhabited_left = NINHAB;
    for (unsigned i = 0; i < NQUADS; ++i) {
	for (unsigned j = 0; j < NQUADS; ++j) {
	    struct quad* q = &Quad[i][j];
	    q->bases = 0;
	    q->pirates = 0;
	    q->stars = nrand(QUAD_STARS);
	    q->holes = nrand(QUAD_HOLES);

	    // NINHAB of NQUADS*NQUADS are inhabited
	    if (!nrand(NQUADS*NQUADS/NINHAB) && inhabited_left)
		q->systemname = --inhabited_left;
	    else
		q->systemname = 0;
	}
    }

    // position starbases
    Now.bases = 2 + nrand(Param.bases-2);
    for (unsigned i = 0; i < Now.bases; ++i) {
	struct quad* q;
	unsigned ix, iy;
	for (;;) {
	    ix = nrand(NQUADS);
	    iy = nrand(NQUADS);
	    q = &Quad[iy][ix];
	    if (q->bases > 0)
		continue;
	    break;
	}
	q->bases = 1;
	Now.base[i].x = ix;
	Now.base[i].y = iy;
    }

    // start at random starbase
    Ship.quad = Now.base [nrand (Now.bases)];

    // position pirates
    unsigned npirates = Param.pirates/2 + nrand(Param.pirates/2);
    Now.time = Now.resource / npirates;
    for (unsigned i = npirates; i > 0;) {
	unsigned plump = nrand(4) + 1;
	if (plump > i)
	    plump = i;
	while (1) {
	    unsigned ix = nrand(NQUADS);
	    unsigned iy = nrand(NQUADS);
	    struct quad* q = &Quad[iy][ix];
	    if (q->pirates + plump > QUAD_PIRATES)
		continue;
	    q->pirates += plump;
	    i -= plump;
	    break;
	}
    }

    // initialize this quadrant
    printf ("%u pirates\n%u starbases ", pirates_remaining(), Now.bases);
    for (unsigned i = 0; i < Now.bases; ++i)
	printf ("%s " QUAD_FMT, i ? "," : "at", Now.base[i].x, Now.base[i].y);
    printf ("\n");
    initquad (false);
    srscan(0);
    attack(0);
}

// Generate quadrant upon entering
//
// A quadrant is initialized from the information held in the Quad
// matrix. Basically, everything is just initialized randomly, except
// for the starship, which goes into a fixed sector.
//
// If there are pirates in the quadrant, the captain is informed that
// the condition is RED, and he is given a chance to put his shields
// up if the computer is working.
//
// The flag 'docked' is set if the ship will be docked on arrival.
//
void initquad (bool docked)
{
    struct quad* q = current_quad();

    // ignored supernova'ed quadrants (this is checked again later anyway
    if (q->stars == SUPERNOVA)
	return;

    // have we blundered into a battle zone w/ shields down?
    if (q->pirates > 0 && !docked && Ship.cond != RED && Ship.cond != CLOAK) {
	printf ("Condition RED\n");
	Ship.cond = RED;
	if (!device_damaged (COMPUTER))
	    shield(1);
    }

    // randomly place in-quadrant entities
    for (unsigned i = 0; i < current_quad()->pirates; ++i) {
	Etc.pirate[i].sect = random_empty_sector();
	Etc.pirate[i].power = Param.piratepwr;
    }
    sort_pirates();
    if (q->bases > 0)
	Etc.starbase = random_empty_sector();
    if (q->systemname != 0)
	Etc.inhabited = random_empty_sector();
    for (unsigned i = 0; i < q->stars; ++i)
	Etc.stars[i] = random_empty_sector();
    for (unsigned i = 0; i < q->holes; ++i)
	Etc.blackholes[i] = random_empty_sector();

    Move.newquad = 1;
}

struct xy random_empty_sector (void)
{
    struct xy r;
    do {
	r.x = nrand(NSECTS);
	r.y = nrand(NSECTS);
    } while (sector_contents(r.x,r.y) != EMPTY);
    return r;
}

//}}}-------------------------------------------------------------------
//{{{ play

// INSTRUCTION READ AND MAIN PLAY LOOP
//
// Well folks, this is it. Here we have the guts of the game.
// This routine executes moves. It sets up per-move variables,
// gets the command, and executes the command. After the command,
// it calls events() to use up time, attack() to have pirates
// attack if the move was not free, and checkcond() to check up
// on how we are doing after the move.

static void save_game (int f UNUSED)
{
    if (save())
	exit (EXIT_SUCCESS);
    puts ("Error: failed to save the game");
}

static void myreset (int v UNUSED)
    { exit (EXIT_SUCCESS); }

static void play (void)
{
    //{{{2 Command table
    static const struct cvntab Comtab[] = {
	{"ca","pture",	capture,	0},
	{"cl","oak",	shield,		-1},
	{"da","mages",	damage_report,	0},
	{"destruct","",	destruct,	0},
	{"do","ck",	dock,		0},
	{"help", "",	help,		0},
	{"i","mpulse",	impulse,	0},
	{"l","rscan",	lrscan,		0},
	{"m","ove",	dowarp,		0},
	{"p","lasers",	plasers,	0},
	{"quit","",	myreset,	0},
	{"ram","",	dowarp,		1},
	{"r","est",	rest,		0},
	{"save","",	save_game,	0},
	{"sh","ield",	shield,		0},
	{"s","rscan",	srscan,		0},
	{"t","orpedo",	torped,		0},
	{"u","ndock",	undock,		0},
	{"v","isual",	visual,		0},
	{"w","arp",	setwarp,	0},
	{}
    };
    //}}}2
    for (;;) {
	Move.free = 1;
	Move.time = 0.0;
	Move.shldchg = 0;
	Move.newquad = 0;
	Move.resting = 0;
	const struct cvntab* r = getcodpar ("Command", Comtab);
	if (!r)
	    continue;
	(*r->value)(r->value2);
	events (0);
	attack (0);
	checkcond();
    }
}

// Check for Condition After a Move
//
// Various ship conditions are checked. First we check
// to see if we have already lost the game, due to running
// out of life support reserves, running out of energy,
// or running out of crew members. The check for running
// out of time is in events().
//
// If we are in automatic override mode (SUPERNOVA && Ship.cond == RED), we
// don't want to do anything else, lest we call autover recursively.
//
// In the normal case, if there is a supernova, we call
// autover() to help us escape. If after calling autover()
// we are still in the grips of a supernova, we get burnt up.
//
// If there are no pirates in this quadrant, we nullify any
// distress calls which might exist.
//
// We then set the condition code, based on the energy level
// and battle conditions.
//
void checkcond (void)
{
    // See if we are still alive and well
    if (Ship.reserves < 0.0)
	lose (L_NOLIFE);
    if (Ship.energy <= 0)
	lose (L_NOENGY);
    if (Ship.crew <= 0)
	lose (L_NOCREW);

    // Call in automatic override if appropriate
    if (current_quad()->stars == SUPERNOVA && Ship.cond != RED)
	autover();
    if (current_quad()->stars == SUPERNOVA)
	lose (L_SNOVA);

    // Cancel distress call if appropriate
    if (!current_quad()->pirates)
	kill_distress_call (Ship.quad.x, Ship.quad.y, 1);

    // set condition code
    if (Ship.cond == DOCKED)
	return;
    if (current_quad()->pirates > 0)
	Ship.cond = RED;
    else if (Ship.energy < Param.energylow)
	Ship.cond = YELLOW;
    else
	Ship.cond = GREEN;
    return;
}

//}}}-------------------------------------------------------------------
//{{{ Game save and restore

//{{{2 s_save_array

#define WVAR(v)	{ &v, sizeof(v) }
#define WARR(v)	{ &v[0], sizeof(v) }

static struct iovec s_save_array[] = {
    WVAR (Ship),
    WVAR (Now),
    WVAR (Etc),
    WVAR (Game),
    WVAR (Move),
    WARR (Quad),
    WARR (Event)
};
//}}}2

static uint16_t sum_save_array (void)
{
    uint16_t sum = 0;
    for (unsigned i = 0; i < ArraySize(s_save_array); ++i)
	sum = bsdsum ((const uint8_t*) s_save_array[i].iov_base, s_save_array[i].iov_len, sum);
    return sum;
}

struct save_header { char magictext[6]; uint16_t sum; };

bool save (void)
{
    char savename [PATH_MAX];
    player_saved_game_dir (ArrayBlock(savename));
    if (0 != access (savename, R_OK))
	mkpath (savename, S_IRWXU);
    if (0 != access (savename, W_OK)) {
	printf ("Error: you are not allowed to write to '%s'\n", savename);
	return false;
    }
    player_saved_game_file (ArrayBlock(savename), SPIRHUNT_SAVE_NAME);

    int fd = creat (savename, S_IRUSR| S_IWUSR);
    if (fd < 0) {
	printf ("Error: unable to create save file '%s': %s\n", savename, strerror(errno));
	return false;
    }
    struct save_header header = {{'s','p','i','r','h','t'}, sum_save_array()};
    if (sizeof(header) != write (fd, &header, sizeof(header))
	|| 0 >= writev (fd, s_save_array, ArraySize(s_save_array))
	|| 0 > close (fd)) {
	printf ("Error writing save file '%s': %s\n", savename, strerror(errno));
	close (fd);
	return false;
    }
    return true;
}

static bool verify_locations (void)
{
    return true;
}

static bool restore (void)
{
    char savename [PATH_MAX];
    player_saved_game_file (ArrayBlock(savename), SPIRHUNT_SAVE_NAME);
    int fd = open (savename, O_RDONLY);
    if (fd < 0)
	return false;

    struct save_header header;
    if (sizeof(header) != read (fd, &header, sizeof(header))
	|| 0 >= readv (fd, s_save_array, ArraySize(s_save_array))) {
	printf ("Error: reading '%s': %s\n", savename, strerror(errno));
	close (fd);
	exit (EXIT_FAILURE);
    }
    close (fd);
    if (memcmp (header.magictext, "spirht", sizeof(header.magictext)) != 0
	|| header.sum != sum_save_array()
	|| !verify_locations()) {
	printf ("Error: saved game '%s' is corrupt. Please delete it.\n", savename);
	exit (EXIT_FAILURE);
    }
    unlink (savename);
    return true;
}

//}}}-------------------------------------------------------------------
//{{{ Score

// PRINT OUT THE CURRENT SCORE
long score (void)
{
    printf ("\n*** Your score:\n");
    int u = Game.pirates_killed;
    int t = u * Param.piratepwr / 4;
    long s = t;
    if (t)
	printf ("%d pirates killed\t\t\t%6d\n", u, t);
    float r = Now.date - Param.date;
    if (r < 1.0)
	r = 1.0;
    r = Game.pirates_killed / r;
    s += (t = 400 * r);
    if (t != 0)
	printf ("Kill rate %.2f pirates/day\t\t%6d\n", r, t);
    r = pirates_remaining();
    r /= Game.pirates_killed + 1;
    s += (t = -400 * r);
    if (t != 0)
	printf ("Penalty for %d pirates remaining\t%6d\n", pirates_remaining(), t);
    if (Move.endgame) {
	s += (t = 100 * (u = 2));
	printf ("Bonus for winning the game\t\t%6d\n", t);
    }
    if (Game.killed) {
	s -= 500;
	printf ("Penalty for getting killed\t\t  -500\n");
    }
    s += (t = -100 * (u = Game.bases_killed));
    if (t != 0)
	printf ("%d starbases killed\t\t\t%6d\n", u, t);
    s += (t = -100 * (u = Game.helps));
    if (t != 0)
	printf ("%d calls for help\t\t\t%6d\n", u, t);
    s += (t = -5 * (u = Game.stars_killed));
    if (t != 0)
	printf ("%d stars destroyed\t\t\t%6d\n", u, t);
    s += (t = -150 * (u = Game.killinhab));
    if (t != 0)
	printf ("%d inhabited starsystems destroyed\t%6d\n", u, t);
    s += (t = 3 * (u = Game.captives));
    if (t != 0)
	printf ("%d pirates captured\t\t\t%6d\n", u, t);
    s += (t = -(u = Game.deaths));
    if (t != 0)
	printf ("%d casualties\t\t\t\t%6d\n", u, t);
    printf ("\n***  TOTAL\t\t\t%14ld\n", s);
    return s;
}

//}}}-------------------------------------------------------------------
//{{{ Win and lose

// Signal game won
//
// This routine prints out the win message, arranges to print out
// your score, tells you if you have a promotion coming to you,
// cleans up the current input line, and exits.
//
// Pretty straightforward, although the promotion algorithm is
// pretty off the wall.

void win (void)
{
    printf ("\nCongratulations, you have saved the Galaxy\n");
    Move.endgame = 1;

    // print and return the score
    score();

    // clean out input, and request new game
    exit (EXIT_SUCCESS);
}

// PRINT OUT LOSER MESSAGES
//
// The messages are printed out, the score is computed and
// printed, and the game is restarted. Oh yeh, any special
// actions which need be taken are taken.

_Noreturn void lose (enum LoseReason why)
{
    Game.killed = (why != L_NOTIME);
    static const char Losemsg[] = {
	"You got bored\0"
	"You ran out of time\0"
	"You ran out of energy\0"
	"You have been destroyed\0"
	"You destroyed yourself by nova'ing that star\0"
	"You have been caught in a supernova\0"
	"You just suffocated in outer space\0"
	"You could not be rematerialized\0"
	"Your ship was ticketed for speeding\0"
	"You have burned up in a star\0"
	"Well, you destroyed yourself, but it didn't do any good\0"
	"You have been captured by pirates and mercilessly tortured\0"
	"Your last crew member died"
    };
    printf ("\n%s\n", zstrn (Losemsg, why, 14));
    score();
    exit (EXIT_SUCCESS);
}

//}}}-------------------------------------------------------------------
//{{{ Miscellaneous

float fnrand (void)
{
    float t = rand() & 077777;
    return t / 32767.0;
}

//}}}-------------------------------------------------------------------
